parse_jenkins_builds.py

```diff
@@ -149,25 +150,79 @@ def grep(filename, pattern, verbose=False):
+    config = json.load(f)
+
 jenkins_queue = dict()
 current_time = get_current_time()
 for element in queue_json["items"]:
-    payload = dict()
-
     job_name = element["task"]["name"]
     queue_id = int(element["id"])
     queue_time = int(element["inQueueSince"])
     labels = element["why"].encode("ascii", "ignore").decode("ascii", "ignore")
     reason = process_queue_reason(labels)
 
-    payload["jenkins_server"] = JENKINS_PREFIX
-    payload["in_queue_since"] = queue_time
-    payload["queue_id"] = queue_id
-    payload["job_name"] = job_name
-    payload["node_labels"] = reason
-    payload["in_queue"] = 1
-    payload["wait_time"] = current_time - queue_time
-    payload["start_time"] = 0
+    payload = {
+        "jenkins_server": JENKINS_PREFIX,
+        "in_queue_since": queue_time,
+        "queue_id": queue_id,
+        "job_name": job_name,
+        "node_labels": reason,
+        "in_queue": 1,
+        "wait_time": current_time - queue_time,
+        "start_time": 0,
+    }
+
+    kill_index = 0
+
+    # Abort stuck jobs
+    if (
+        job_name in config["whitelist"]
+        and reason.endswith("-offline")
+        and reason != "multiple-offline"
+        and (payload["wait_time"] / 1000 > config["custom"].get(job_name, config["timeout"]))
+    ):
+        params = dict(
+            line.split("=", 1) for line in element["params"].strip().splitlines() if "=" in line
+        )
+
+        if "rocm" not in (params.get("GPU_FLAVOR"), params.get("TEST_FLAVOR")):
+            continue
+
+        # Try to reconnect the node
+        node_name = reason.rsplit("-", 1)[0]
+        connect_node = os.environ.get("JENKINS_CLI_CMD") + " connect-node " + node_name + " -f"
+        try_count = 1
+        while try_count < 4:
+            try_count += 1
+            ret = os.system(connect_node)
+            if ret == 0:
+                break
+
+        if try_count == 4:
+            try:
+                pull_request = params["PULL_REQUEST"]
+                main_params = f"PULL_REQUEST={pull_request}"
+                release = params["RELEASE_FORMAT"]
+                context = params["CONTEXT_PREFIX"]
+                upload_unique_id = params["UPLOAD_UNIQ_ID"]
+                test_flavor = params["TEST_FLAVOR"]
+            except KeyError:
+                continue
+
+            other_params = ";".join(f"{k}={v}" for k, v in params if k != "PULL_REQUEST")
+
+            with open(f"abort-{kill_index}.prop", "w") as f:
+                f.write(f"UPLOAD_UNIQ_ID={upload_unique_id}\n")
+                f.write(f"PULL_REQUEST={pull_request}\n")
+                f.write(f"CONTEXT={context}\n")
+                f.write(f"JENKINS_PROJECT_TO_KILL={job_name}\n")
+                f.write(f"JENKINS_PROJECT_PARAMS={main_params}\n")
+                f.write(f"EXTRA_PARAMS={other_params}\n")
+                f.write(f"RELEASE_FORMAT={release}\n")
+                f.write(f"TEST_FLAVOR={test_flavor}\n")
+
+            kill_index += 1
```

Comment:
@iarspider , I wuld suggest tomove these functions to `_helper_functions.sh` file. `_helper_functions.sh` is already sources in `run-pr-relvals.sh and run-pr-unittests.sh` via `setup-pr-test-env.sh`